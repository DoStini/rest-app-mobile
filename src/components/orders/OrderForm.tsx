import * as Yup from "yup";
import theme from "../../theme";
import Button from "../Button";
import { Formik } from "formik";
import { ScrollView, TextInput, View } from "react-native";
import Text from "../Text";
import React from "react";
import RadioButtonGroup, { RadioButtonItem } from "expo-radio-button";
import Divider from "../Divider";

type OrderFormProps = {
  initialValues: { name: string; tableId: number };
  handleSubmit: (values: { name: string; tableId: number }) => void;
  tables: { name: string; id: number }[];
  loading: boolean;
};

const FormValidationSchema = Yup.object().shape({
  tableId: Yup.number().notOneOf([0], "Table is required"),
  name: Yup.string().required("Name is required"),
});

const OrderForm = ({
  initialValues,
  handleSubmit,
  tables,
  loading,
}: OrderFormProps) => {
  return (
    <ScrollView
      automaticallyAdjustKeyboardInsets={true}
      style={{ padding: 0, margin: 0 }}
    >
      <Formik
        initialValues={initialValues}
        validationSchema={FormValidationSchema}
        onSubmit={handleSubmit}
      >
        {({ handleSubmit, errors, values, setFieldValue }) => (
          <View>
            <Text fontSize="body" fontWeight="bold" color="textPrimary">
              Table
            </Text>

            <RadioButtonGroup
              onSelected={(value: string) => setFieldValue("tableId", value)}
              selected={values.tableId}
              size={15}
              containerStyle={{ marginVertical: 10 }}
              radioBackground={theme.colors.selectedColor}
            >
              {tables?.map((table) => (
                <RadioButtonItem
                  label={table.name}
                  value={table.id}
                  key={table.id}
                  style={{ marginVertical: 7, fontSize: theme.fontSizes.body }}
                />
              ))}
            </RadioButtonGroup>

            {errors.tableId && (
              <Text fontSize="small" color="error">
                {errors.tableId}
              </Text>
            )}

            <Divider />

            <Text fontSize="body" fontWeight="bold" color="textPrimary">
              Table's Name
            </Text>

            <TextInput
              placeholder="Table's Name"
              style={{}}
              onChangeText={(text) => setFieldValue("name", text)}
              value={values.name}
            />

            {errors.name && (
              <Text fontSize="small" color="error">
                {errors.name}
              </Text>
            )}

            <Divider />

            <Button
              text="Create Order"
              onPress={handleSubmit}
              loading={loading}
            />
          </View>
        )}
      </Formik>
    </ScrollView>
  );
};

export default OrderForm;
