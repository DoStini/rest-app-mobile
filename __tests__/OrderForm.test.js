import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import OrderForm from "../src/components/orders/OrderForm";

const initialValues = { name: "", tableId: 0 };
const handleSubmit = jest.fn();
const tables = [
  { name: "Table 1", id: 1 },
  { name: "Table 2", id: 2 },
];
const loading = false;
const submitText = "Submit";

describe("OrderForm component", () => {
  it("renders correctly", () => {
    const { getByText, getByPlaceholderText } = render(
      <OrderForm
        initialValues={initialValues}
        handleSubmit={handleSubmit}
        tables={tables}
        loading={loading}
        submitText={submitText}
      />
    );
    expect(getByText("Table 1")).toBeTruthy();
    expect(getByText("Table 2")).toBeTruthy();

    expect(getByPlaceholderText("Table's Name")).toBeTruthy();
    expect(getByText(submitText)).toBeTruthy();
  });

  it("Snapshot is correct", () => {
    const { toJSON } = render(
      <OrderForm
        initialValues={initialValues}
        handleSubmit={handleSubmit}
        tables={tables}
        loading={loading}
        submitText={submitText}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("requires necessary fields", async () => {
    const { getByTestId, findByText } = render(
      <OrderForm
        initialValues={initialValues}
        handleSubmit={handleSubmit}
        tables={tables}
        loading={loading}
        submitText={submitText}
      />
    );
    fireEvent.press(getByTestId("orderFormButton"));

    expect(await findByText("Name is required")).toBeTruthy();
    expect(await findByText("Table is required")).toBeTruthy();
  });

  /*it("handles user input correctly", async () => {
    const { getByTestId, getByText, getByPlaceholderText } = render(
      <OrderForm
        initialValues={initialValues}
        handleSubmit={handleSubmit}
        tables={tables}
        loading={loading}
        submitText={submitText}
      />
    );

    fireEvent.press(getByText("Table 1"));
    fireEvent.changeText(getByPlaceholderText("Table's Name"), "Test Name");
    fireEvent.press(getByTestId("orderFormButton"));

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledWith({
        name: "Test Name",
        tableId: 1,
      });
    });
  });*/
});
