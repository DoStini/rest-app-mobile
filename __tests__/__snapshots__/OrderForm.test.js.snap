// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm component Snapshot is correct 1`] = `
<RCTScrollView
  automaticallyAdjustKeyboardInsets={true}
  style={
    {
      "margin": 0,
      "padding": 0,
    }
  }
>
  <View>
    <View>
      <Text
        style={
          [
            {
              "color": "black",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "400",
            },
            {
              "color": "black",
            },
            {
              "fontSize": 16,
            },
            {
              "fontWeight": "700",
            },
            undefined,
            undefined,
          ]
        }
      >
        Table
      </Text>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                {
                  "borderColor": "#CCC",
                  "borderRadius": 50,
                  "borderWidth": 2,
                  "fontSize": 16,
                  "height": 15,
                  "marginVertical": 7,
                  "padding": 2,
                  "width": 15,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "rgba(0,0,0,0)",
                    "borderRadius": 50,
                    "height": "100%",
                    "width": "100%",
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "marginLeft": 3,
                }
              }
            >
              Table 1
            </Text>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                {
                  "borderColor": "#CCC",
                  "borderRadius": 50,
                  "borderWidth": 2,
                  "fontSize": 16,
                  "height": 15,
                  "marginVertical": 7,
                  "padding": 2,
                  "width": 15,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "rgba(0,0,0,0)",
                    "borderRadius": 50,
                    "height": "100%",
                    "width": "100%",
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "marginLeft": 3,
                }
              }
            >
              Table 2
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "borderBottomColor": "#bbbbbbbb",
              "borderBottomWidth": 2,
              "borderTopColor": "#bbbbbbbb",
              "borderTopWidth": 0,
              "marginHorizontal": -20,
              "marginVertical": 10,
            },
            undefined,
          ]
        }
      />
      <Text
        style={
          [
            {
              "color": "black",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "400",
            },
            {
              "color": "black",
            },
            {
              "fontSize": 16,
            },
            {
              "fontWeight": "700",
            },
            undefined,
            undefined,
          ]
        }
      >
        Table's Name
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Table's Name"
        style={{}}
        value=""
      />
      <View
        style={
          [
            {
              "borderBottomColor": "#bbbbbbbb",
              "borderBottomWidth": 2,
              "borderTopColor": "#bbbbbbbb",
              "borderTopWidth": 0,
              "marginHorizontal": -20,
              "marginVertical": 10,
            },
            undefined,
          ]
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        animatedStyle={
          {
            "value": {},
          }
        }
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#96744E",
            "opacity": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 10,
          }
        }
        testID="orderFormButton"
      >
        <View
          style={
            {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 16,
              }
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
